src/Nethermind/Nethermind.Consensus.AuRa/Transactions/TxPermissionFilter.cs

```diff
@@ -0,0 +1,118 @@
+// 
+
+using System;
+using Nethermind.Consensus.AuRa.Contracts;
+using Nethermind.Core;
+using Nethermind.Core.Caching;
+using Nethermind.Core.Crypto;
+using Nethermind.Dirichlet.Numerics;
+using Nethermind.Logging;
+
+namespace Nethermind.Consensus.AuRa.Transactions
+{
+    public class TxPermissionFilter : ITxPermissionFilter
+    {
+        private readonly TransactionPermissionContract _contract;
+        private readonly ITxPermissionFilter.Cache _cache;
+        private readonly ILogger _logger;
+        
+        public TxPermissionFilter(TransactionPermissionContract contract, ITxPermissionFilter.Cache cache, ILogManager logManager)
+        {
+            _contract = contract ?? throw new ArgumentNullException(nameof(contract));
+            _cache = cache ?? throw new ArgumentNullException(nameof(cache));
+            _logger = logManager?.GetClassLogger<TxPermissionFilter>() ?? throw new ArgumentNullException(nameof(logManager));
+        }
+        
+        public bool IsAllowed(Transaction tx, BlockHeader blockHeader)
+        {
+            if (blockHeader.Number + 1 < _contract.ActivationBlock)
+            {
+                return true;
+            }
+            else
+            {
+                var txType = GetTxType(tx);
+                var txPermissions = GetPermissions(tx, blockHeader);
+                if (_logger.IsTrace) _logger.Trace($"Given transaction data: sender: {tx.SenderAddress} to: {tx.To} value: {tx.Value}, gas_price: {tx.GasPrice}. Permissions required: {txType}, got: {txPermissions}.");
+                return (txPermissions & txType) == txType;
+            }
+        }
+
+        private TransactionPermissionContract.TxPermissions GetPermissions(Transaction tx, BlockHeader blockHeader)
+        {
+            var key = (blockHeader.Hash, tx.SenderAddress);
+            var txCachedPermissions = _cache.Permissions.Get(key);
+            return txCachedPermissions ?? GetPermissionsFromContract(tx, blockHeader, key);
+        }
+
+        private TransactionPermissionContract.TxPermissions GetPermissionsFromContract(Transaction tx, BlockHeader blockHeader, in (Keccak Hash, Address SenderAddress) key)
+        {
+            TransactionPermissionContract.TxPermissions txPermissions = TransactionPermissionContract.TxPermissions.None;
+            bool cache = true;
+            
+            var versionedContract = GetVersionedContract(blockHeader);
+            if (versionedContract == null)
+            {
+                if (_logger.IsError) _logger.Error("Unknown version of tx permissions contract is used.");
+            }
+            else
+            {
+                if (_logger.IsTrace) _logger.Trace($"Version of tx permission contract: {versionedContract.Version}.");
+                
+                try
+                {
+                    (txPermissions, cache) = versionedContract.AllowedTxTypes(blockHeader, tx);
+                }
+                catch (AuRaException e)
```

Comment:
what happens on other exceptions?

Reply:
We allow them to bubble up. This one is the only expected we know we can handle.